
#' Read VCF to R
#'
#' @description
#' This function reads the VCF file into R and converts it to a data.table format. While genomic range is given, [HTSlib](https://github.com/samtools/htslib), a wrapped C library, reads directly from a region instead of loading the full VCF file to memory. It is mandatory to compress the VCF file using bgzip and create an index using tabix.
#'
#'
#' @param file The VCF file.
#' @param range To specify the range you want to read from the genome. e.g. "1:1000-100000" (genomic range), "1:12345"(single variant), "1:12345-12345" (single variant), NULL (all variants)
#' @import data.table
#' @import Rsamtools
#' @import stringr
#' @importFrom GenomicRanges GRanges
#' @importFrom IRanges IRanges
#' @return A list which include a genotype matrix and a marker information data.table. Return NULL if no variant found withing given range.
#'
#' @export
read_vcf = function(file, range = NULL){

  ### Check if vcf file and its index exist ###
  if(! file.exists(file)){
    stop("VCF file not found.")
  }
  if(! file.exists(paste0(file, ".tbi"))){
    stop("Index file not found. The index file should be generated by tabix and placed in the same folder as the vcf file.")
  }

  ### Load VCF header ###
  vcfHeader = c("CHROM", "POS", "ID", "REF", "ALT", "QUAL", "FILTER", "INFO", "FORMAT", as.character(Rsamtools::scanBcfHeader(file)[[file]]$Sample))

  ### Read the whole VCF file if range is not provided ###
  if(is.null(range)){
    tmp = data.table::fread(file)
    colnames(tmp) = vcfHeader
  }


  ### If only range provided ###
  if(! is.null(range)){
    # Prepare range information
    splitRange = stringr::str_split(range, pattern = ":|-", simplify = T, n=3)[1, ]
    chr = splitRange[1]
    start = as.numeric(splitRange[2])
    end = ifelse(splitRange[3]=="", as.numeric(splitRange[2]), as.numeric(splitRange[3]))
    width = end-start+1

    # Load file
    tmp = unlist(Rsamtools::scanTabix(file, param=GenomicRanges::GRanges(seqnames = chr, IRanges::IRanges(start, end, width))))
    if(length(tmp) == 1){
      tmp = data.table::as.data.table(t(unlist(stringr::str_split(tmp, "\t"))))
      colnames(tmp) = vcfHeader
    }else if (length(tmp) > 1){
      tmp = data.table::fread(text = tmp)
      colnames(tmp) = vcfHeader
    }
  }


  ### Reformat and return result
  if(is.null(dim(tmp))){
    vcf = NULL
  }else{
    marker = tmp[,1:5]
    oldGeno = t(as.matrix(tmp[, -1:-9]))
    geno = matrix(NA, nrow(oldGeno), ncol(oldGeno))
    geno[stringr::str_detect(oldGeno, "(0/0)|(0|0)")] = 0
    geno[stringr::str_detect(oldGeno, "(0/1)|(1/0)|(0|1)|(1|0)")] = 1
    geno[stringr::str_detect(oldGeno, "(1/1)|(1|1)")] = 2
    rownames(geno) = vcfHeader[-1:-9]
    colnames(geno) = marker$ID
    vcf = list(geno=geno, marker=marker)
  }

  return(vcf)
}

